src/env/Boundary.cpp

```diff
@@ -0,0 +1,265 @@
+            if (dist < 1.1225 * sigma) {
+
+                const double inv_dist = sigma / dist;
+                const double inv_dist_2 = inv_dist * inv_dist;       // (sigma / dist)^2
+                const double inv_dist_6 = inv_dist_2 * inv_dist_2 * inv_dist_2; // (sigma / dist)^6
+                const double inv_dist_12 = inv_dist_6 * inv_dist_6; // (sigma / dist)^12
+
+                return - 24 * epsilon * (inv_dist_6 - 2 * inv_dist_12) / dist;
+            }
+            return 0.0;
+        };
+    }
+
+    BoundaryForce Boundary::InverseDistanceForce(const double cutoff, const double pre_factor, const int exponent) {
+        return [=](const double dist) {
+            if (dist < cutoff) return pre_factor/pow(dist, exponent);
+            return 0.0;
+        };
+    }
+
+    bool Boundary::requires_force_function() const {
+        for (auto rule : rules) {
+            if (rule == REPULSIVE_FORCE) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    bool Boundary::has_force_function() const {
+        return static_cast<bool>(force);
+    }
+
+
+    void Boundary::apply_rule(const int3& normal, Particle& particle, const GridCell& current_cell) const {
+        switch (rules[face_normal_to_idx(normal)]) {
+            case OUTFLOW: outflow_rule(particle, current_cell); break;
+            case PERIODIC: periodic_rule(particle, normal, current_cell); break;
+            case REPULSIVE_FORCE: repulsive_force_rule(particle, normal, current_cell); break;
+            case VELOCITY_REFLECTION: velocity_reflection_rule(particle, normal, current_cell); break;
+        }
+    }
+
+
+    void Boundary::outflow_rule(Particle& particle, const GridCell& previous_cell) const {
+        // this rule may be called when applying all boundary conditions of a cell, hence we need a check if
+        // the particle is outside or inside
+        if (previous_cell.type == GridCell::OUTSIDE) {
+            particle.state = Particle::DEAD;
+            particle.update_grid();
+        }
+    }
+
+    void Boundary::periodic_rule(Particle& particle,const int3& normal, const GridCell& current_cell) const {
+        if (current_cell.type == GridCell::OUTSIDE) {
+            int axis = axis_from_normal(normal);
+            vec3 dx = {0,0,0};
+            if (normal[axis] == 1) dx[axis] = - extent[axis];
+            else if (normal[axis] == -1) dx[axis] = extent[axis];
+            particle.update_position(dx);
+            particle.update_grid();
+        }
+    }
+
+    void Boundary::repulsive_force_rule(Particle& particle, const int3 & normal, const GridCell& cell) const {
+        const double dist = distance_to_boundary(particle, normal, cell);
+
+        if (dist != 0) { // particles may be initialized at the boundary
+            vec3 df = {};
+            int axis = axis_from_normal(normal);
+            df[axis] = force(dist) * normal[axis];
+            particle.force = particle.force - df;
```

Comment:
we chose do not include a check here, since the user can specify the repulsive force how they want to using a lambda function and the set_boundary_force method. see for instance the static method Boundary::LennardJonesForce, which is one of the forces we pre-define for a user ready to use.

Reply:
But this has to be ensured somewhere. In context of our simulation, this definitely has to be checked.

But even if you want to keep it general, you could still just use something like `std::max( 0, calc_force)`, so that your ensure that you never apply attractive force!